proposals/factory_constructors.md

```diff
@@ -0,0 +1,216 @@
+    ;
+
+factory_constructor_modifier
+    : 'new'
+    | 'public'
+    | 'protected'
+    | 'internal'
+    | 'private'
+    | 'static'
+    | 'virtual'
+    | 'sealed'
+    | 'override'
+    | 'abstract'
+    | factory_constructor_modifier_unsafe
+    ;
+
+factory_constructor_modifier_unsafe
+    : 'unsafe'
+    ;
+
+factory_constructor_declarator
+    : type '(' formal_parameter_list? ')'
+    ;
+
+factory_constructor_body
+    : block
+    | ';'
+    ;
+```
+
+Some examples:
+
+```cs
+public static interface I
+{
+    int MyProperty { get; init; }
+    // Define a copy constructor for the ability to `with` any `I` implementation
+    factory I();
+}
+public abstract class AbsI : I
+{
+    public int MyProperty { get; init; }
+
+    protected abstract new factory I();
+    factory I.I()
+        => new this.I() // `this.I` is required to disabmiguate from the default parameterless constructor
+           {
+                MyField = this.MyField
+           };
+}
+public abstract class DerivedI : AbsI
+{
+    public int DerivedProperty { get; set; }
+    protected override factory I() => new DerivedProperty() { DerivedProperty = this.DerivedProperty };
+}
+
+public class IFactory
+{
+    public factory I(int param1, int param2) => ...;
+    public factory I(string param1) => ...;
+    public static factory I(object param1) => ...;
+}
+```
+
+Factory constructors are invoked by calling `new`, and providing a primary expression that resolves to a factory constructor.
+
+```antlr
+object_creation_expression
+    : 'new' primary_expression_or_type '(' argument_list? ')' object_or_collection_initializer?
+    | 'new' primary_expression_or_type object_or_collection_initializer
+    ;
+
+primary_expression_or_type
+    : primary_expression
+    | type
+    ;
+```
+
+For an object creation expression, the compiler will first attempt to bind the `primary_expression_or_type` as a type. If it binds,
+and if overload resolution with the provided arguments returns anything other than an empty set, then no attempt is made to bind
+the `primary_expression_or_type` as a `primary_expression`. However, if it fails to bind as a type or overload resolution fails to
+find an applicable constructor, the compiler will bind the `primary_expression_or_type` as a `primary_expression`, and attempt to
+resolve it as a method group. If overload resolution returns a factory constructor, then that is the constructor that is used by the
+object creation expression.
+
+Examples:
+
+```cs
+var iFactory = new IFactory();
+var i1 = new iFactory.I(1, 2) { MyProperty = 3 };
+var i2 = new iFactory.I("hello world!");
+var i3 = new (new IFactory.I(null)).I();
+```
+
+Factory constructors have the following restrictions:
+
+* The return expression must be
+  * new expression including object / collection initializers
+  * Call to a member annotated with [Factory]
+  * A struct value
+  * A ternary or collascing expression where both branches meet the above requirements
```

Comment:
or switch expression?

Reply:
Jared and Andy had concerns about stack spilling and verifiability (which is why it was omitted from Jared's original proposal, where I copied this list from). We'll have to work through it but I do certainly want to see them supported here.